contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,758 @@
+      // (we subtract g//64 because of EIP-150)
+      if iszero(gt(sub(g, div(g, 64)), gasAmount)) { revert(0, 0) }
+      // solidity calls check that a contract actually exists at the destination, so we do the same
+      if iszero(extcodesize(target)) { revert(0, 0) }
+      // call and return whether we succeeded. ignore return data
+      success := call(gasAmount, target, 0, add(data, 0x20), mload(data), 0, 0)
+    }
+    return success;
+  }
+
+  // Offsets into fulfillRandomnessRequest's proof of various values
+  //
+  // Public key. Skips byte array's length prefix.
+  uint256 private constant PUBLIC_KEY_OFFSET = 0x20;
+  // Seed is 7th word in proof, plus word for length, (6+1)*0x20=0xe0
+  uint256 private constant PRESEED_OFFSET = 7*0x20;
+
+  function fulfillRandomWords(
+    bytes memory proof
+  )
+    external
+  {
+    uint256 startGas = gasleft();
+    (
+      bytes32 keyHash,
+      uint256 requestId,
+      uint256 randomness,
+      FulfillmentParams memory fp
+    ) = getRandomnessFromProof(proof);
+
+
+    uint256[] memory randomWords = new uint256[](fp.numWords);
+    for (uint256 i = 0; i < fp.numWords; i++) {
+      randomWords[i] = uint256(keccak256(abi.encode(randomness, i)));
+    }
+
+    // Prevent re-entrancy. The user callback cannot call fulfillRandomWords again
+    // with the same proof because this getRandomnessFromProof will revert because the requestId
+    // is gone.
+    delete s_commitments[requestId];
+    VRFConsumerBaseV2 v;
+    bytes memory resp = abi.encodeWithSelector(v.rawFulfillRandomWords.selector, requestId, randomWords);
+    uint256 gasPreCallback = gasleft();
+    if (gasPreCallback < fp.callbackGasLimit) {
+      revert InsufficientGasForConsumer(gasPreCallback, fp.callbackGasLimit);
+    }
+    // Call with explicitly the amount of callback gas requested
+    // Important to not let them exhaust the gas budget and avoid oracle payment.
+    bool success = callWithExactGas(fp.callbackGasLimit, fp.sender, resp);
+    emit RandomWordsFulfilled(requestId, randomWords, success);
+
+    // We want to charge users exactly for how much gas they use in their callback.
+    // The gasAfterPaymentCalculation is meant to cover these additional operations where we
+    // decrement the subscription balance and increment the oracles withdrawable balance.
+    // We also add the flat link fee to the payment amount.
+    // Its specified in millionths of link, if s_config.fulfillmentFlatFeeLinkPPM = 1
+    // 1 link / 1e6 = 1e18 juels / 1e6 = 1e12 juels.
+    uint96 payment = calculatePaymentAmount(startGas, s_config.gasAfterPaymentCalculation, s_config.fulfillmentFlatFeeLinkPPM, tx.gasprice);
+    if (s_subscriptions[fp.subId].balance < payment) {
+      revert InsufficientBalance();
+    }
+    s_subscriptions[fp.subId].balance -= payment;
+    s_withdrawableTokens[s_serviceAgreements[keyHash]] += payment;
+  }
+
+  // Get the amount of gas used for fulfillment
+  function calculatePaymentAmount(
```

Comment:
For full unit test coverage, I think we should add tests for private functions like this (especially given the amount of math going on here).

We can do this by creating a test helper contract that inherits this and exposes a function like `calculatePaymentAmountTest(...)` and directly calls this.

Reply:
I think I have to make it `internal` to be able to do that, which appears ok because this function is stateless. Will try it out

Reply:
done